### PR TITLE
Merging the bdf-usecase3 branch to main

### PR DESCRIPTION
**Pull request by vitaglianog**
This branch includes the following updates to the codebase:

- Include the code to run the use case 3 from the BDF project. This can be found in `demos/bdf-usecase3.py` and in `demos/bdf-suite.pdf `
- A new ExecutionEngine for Streaming execution (in `execution/streaming_execution.py`). Using this execution engine enables to split plan generation and plan execution for each input record so that the results can be inspected before the full dataset is processed.
- The Optimizer class now supports non-unique field names within different schemata of a workload, disambiguated through the name of the schema class.
- A library of commonly used UDFs converts can be found in `utils/udfs`.
- Adedd support for the `pypdf` processor to extract text from PDFFiles.

The code also include minor changes and bugfixes picked up during development.

### CODE REVIEW

**Comment by mdr223** [comment]
Taking a look at the PR now! Before I dive deep, one question I've been having is whether we should:
1. continue committing the `testdata/` directory
2. `tar` the testdata directory and commit a single tar file (perhaps using `git lfs`)
3. drop the `testdata` directory, make the files publicly available in google drive, and add instructions for downloading them in the README.md

Curious what others think is the best option for the long term

**Comment by vitaglianog** [comment]
I am not a big fan of option 2. since I have bad experiences with git lfs, it's sometimes a clunky integration. 
So I am in favor of option 3. maybe including a script like `download.sh` that automatically downloads and extracts all files into the correct place, to be run once.

**Comment by mdr223** [comment]
100% agreed re: the clunkiness of `git lfs` and I love the idea of a `download.sh`

**Review by mdr223 — COMMENTED** [review]
Should hopefully see comments now

File: src/palimpzest/operators/convert.py [hunk 1]
```diff
@@ -512,7 +515,12 @@ def __call__(self, candidate: DataRecord) -> List[DataRecordsWithStats]:
 
         # construct list of dictionaries where each dict. has the (field, value) pairs for each generated field
         # list is indexed per record
-        n_records = max([len(lst) for lst in field_answers.values()])
+        try:
+            n_records = max([len(lst) for lst in field_answers.values()])
```
[the remark this answers is not in the document]
Comment (vitaglianog) [reply src/palimpzest/operators/convert.py:519]: Leaving this comment open since I'm not sure to which you refer

**Pull request merged**